ballista/scheduler/src/state/backend/memory.rs

```diff
@@ -0,0 +1,395 @@
+                        .collect::<Vec<(String, Vec<u8>)>>()
+                } else {
+                    space_state
+                        .value()
+                        .iter()
+                        .map(|e| (format!("{}/{}", space_key, e.0), e.1.clone()))
+                        .collect::<Vec<(String, Vec<u8>)>>()
+                }
+            })
+            .unwrap_or_default())
+    }
+
+    async fn scan_keys(&self, keyspace: Keyspace) -> Result<HashSet<String>> {
+        let space_key = Self::get_space_key(&keyspace);
+        Ok(self
+            .states
+            .get(&space_key)
+            .map(|space_state| {
+                space_state
+                    .value()
+                    .iter()
+                    .map(|e| format!("{}/{}", space_key, e.0))
+                    .collect::<HashSet<String>>()
+            })
+            .unwrap_or_default())
+    }
+
+    async fn put(&self, keyspace: Keyspace, key: String, value: Vec<u8>) -> Result<()> {
+        let space_key = Self::get_space_key(&keyspace);
+        if !self.states.contains_key(&space_key) {
+            self.states.insert(space_key.clone(), BTreeMap::default());
+        }
+        self.states
+            .get_mut(&space_key)
+            .unwrap()
+            .value_mut()
+            .insert(key.clone(), value.clone());
+
+        // Notify subscribers
+        let full_key = format!("{}/{}", space_key, key);
+        if let Some(res) = self.subscribers.reserve(&full_key) {
+            let event = WatchEvent::Put(full_key, value);
+            res.complete(&event);
+        }
+
+        Ok(())
+    }
+
+    async fn apply_txn(&self, ops: Vec<(Operation, Keyspace, String)>) -> Result<()> {
+        let mut locks = vec![];
+        for (_, keyspace, key) in ops.iter() {
+            locks.push(self.lock(keyspace.clone(), key).await?);
+        }
+
+        for (op, keyspace, key) in ops.into_iter() {
+            match op {
+                Operation::Delete => {
+                    self.delete(keyspace, &key).await?;
+                }
+                Operation::Put(value) => {
+                    self.put(keyspace, key, value).await?;
+                }
+            };
+        }
+
+        Ok(())
+    }
+
+    async fn mv(
+        &self,
+        from_keyspace: Keyspace,
+        to_keyspace: Keyspace,
+        key: &str,
+    ) -> Result<()> {
+        let from_space_key = Self::get_space_key(&from_keyspace);
+
+        if let Some(from_space_state) = self.states.get(&from_space_key) {
+            if let Some(state) = from_space_state.value().get(key) {
+                let ops = vec![
+                    (Operation::Delete, from_keyspace, key.to_owned()),
+                    (Operation::Put(state.clone()), to_keyspace, key.to_owned()),
+                ];
+                self.apply_txn(ops).await
+            } else {
+                // TODO should this return an error?
```

Comment:
I think this should be consistent with the Sled and Etcd backends which I believe would both return an error in this scenario

Reply:
Actually, it's consistent with the `sled` implementation